internal/core/entity/entity.go

```diff
@@ -0,0 +1,200 @@
+	HTTPFailures int   `json:"http_failures,omitempty"`
+}
+
+type Active struct {
+	Type                   string    `json:"type,omitempty"`
+	Timeout                int       `json:"timeout,omitempty"`
+	Concurrency            int       `json:"concurrency,omitempty"`
+	Host                   string    `json:"host,omitempty"`
+	Port                   int       `json:"port,omitempty"`
+	HTTPPath               string    `json:"http_path,omitempty"`
+	HTTPSVerifyCertificate string    `json:"https_verify_certificate,omitempty"`
+	Healthy                Healthy   `json:"healthy,omitempty"`
+	UnHealthy              UnHealthy `json:"unhealthy,omitempty"`
+	ReqHeaders             []string  `json:"req_headers,omitempty"`
+}
+
+type Passive struct {
+	Type      string    `json:"type,omitempty"`
+	Healthy   Healthy   `json:"healthy,omitempty"`
+	UnHealthy UnHealthy `json:"unhealthy,omitempty"`
+}
+
+type HealthChecker struct {
+	Active  Active  `json:"active,omitempty"`
+	Passive Passive `json:"passive,omitempty"`
+}
+
+type UpstreamTLS struct {
+	ClientCert string `json:"client_cert,omitempty"`
+	ClientKey  string `json:"client_key,omitempty"`
+}
+
+type KeepLivePool struct {
+	Size        int `json:"size,omitempty"`
+	IdleTimeout int `json:"idle_timeout,omitempty"`
+	Requests    int `json:"requests,omitempty"`
+}
+
+type UpstreamArg struct {
+	NamespaceID string `json:"namespace_id,omitempty"`
+	GroupName   string `json:"group_name,omitempty"`
+}
+
+type UpstreamDef struct {
+	Nodes         interface{}       `json:"nodes,omitempty"`
+	Retries       int               `json:"retries,omitempty"`
+	Timeout       *Timeout          `json:"timeout,omitempty"`
+	Type          string            `json:"type,omitempty"`
+	Checks        *HealthChecker    `json:"checks,omitempty"`
+	HashOn        string            `json:"hash_on,omitempty"`
+	Key           string            `json:"key,omitempty"`
+	Scheme        string            `json:"scheme,omitempty"`
+	DiscoveryType string            `json:"discovery_type,omitempty"`
+	DiscoveryArgs *UpstreamArg      `json:"discovery_args,omitempty"`
+	PassHost      string            `json:"pass_host,omitempty"`
+	UpstreamHost  string            `json:"upstream_host,omitempty"`
+	Name          string            `json:"name,omitempty"`
+	Desc          string            `json:"desc,omitempty"`
+	ServiceName   string            `json:"service_name,omitempty"`
+	Labels        map[string]string `json:"labels,omitempty"`
+	TLS           *UpstreamTLS      `json:"tls,omitempty"`
+	Pool          *KeepLivePool     `json:"keepalive_pool,omitempty"`
+}
+
+func (u *UpstreamDef) SetNodes(nodes []*Node) {
+	u.Nodes = NodesFormat(nodes)
+}
+
+type Upstream struct {
+	BaseInfo
+	UpstreamDef
+}
+
+func (u *Upstream) SetNodes(nodes []*Node) {
+	(u.UpstreamDef).SetNodes(nodes)
+}
+
+// --- structures for upstream end ---
+
+type Route struct {
+	BaseInfo
+	URI             string                 `json:"uri,omitempty"`
```

Comment:
Better only keep the required fields.

Reply:
One question I have here is that if we remove the non-required fields, then when we update etcd, the serialization of the struct will also be without the filed value, so will it overwrite the existing value in etcd?

Reply:
Can we find a way to solve it? There will be new fields added across versions. We need to keep the fields that can't be decoded.

Reply:
In the latest code, I tried a way to remove unnecessary fields from the entities.